servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/DefaultPriorityQueue.java

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright © 2021 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.concurrent.api;
+
+import java.util.AbstractQueue;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static io.servicetalk.concurrent.api.PriorityQueueNode.INDEX_NOT_IN_QUEUE;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A priority queue which uses natural ordering of elements. Elements are also required to be of type
+ * {@link PriorityQueueNode} for the purpose of maintaining the index in the priority queue.
+ * @param <T> The object that is maintained in the queue.
+ */
+final class DefaultPriorityQueue<T extends PriorityQueueNode> extends AbstractQueue<T> {
```

Comment:
I implemented this in Netty and borrowed it here. The benefit of this queue vs jdk is constant time removal and location within the array.

Reply:
Pls include the Copyright header for Netty.